txscript/sigcache_test.go

```diff
@@ -0,0 +1,141 @@
+	privKey, err := btcec.NewPrivateKey(btcec.S256())
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	msgHash := make([]byte, wire.HashSize)
+	if _, err := rand.Read(msgHash); err != nil {
+		return nil, nil, nil, err
+	}
+
+	sig, err := privKey.Sign(msgHash)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	return msgHash, sig, privKey.PubKey(), nil
+}
+
+// TestSigCacheAddExists tests the ability to add, and later check the
+// existence of a signature triplet in the signature cache.
+func TestSigCacheAddExists(t *testing.T) {
+	sigCache := NewSigCache(200)
+
+	// Generate a random sigCache entry triplet.
+	msg1, sig1, key1, err := genRandomSig()
+	if err != nil {
+		t.Errorf("unable to generate random signature test data")
+	}
+
+	// Add the triplet to the signature cache.
+	sigCache.Add(msg1, sig1, key1)
+
+	// The previously added triplet should now be found within the sigcache.
+	sig1Copy, _ := btcec.ParseSignature(sig1.Serialize(), btcec.S256())
+	key1Copy, _ := btcec.ParsePubKey(key1.SerializeCompressed(), btcec.S256())
+	if !sigCache.Exists(msg1, sig1Copy, key1Copy) {
+		t.Errorf("previously added item not found in signature cache")
+	}
+}
+
+// TestSigCacheAddEvictEntry tests the eviction case where a new signature
+// triplet is added to a full signature cache which should trigger randomized
+// eviction, followed by adding the new element to the cache.
+func TestSigCacheAddEvictEntry(t *testing.T) {
+	// Create a sigcache that can hold up to 10 entries.
+	sigCacheSize := 100
+	sigCache := NewSigCache(sigCacheSize)
+
+	// Fill the sigcache up with some random sig triplets.
+	for i := 0; i < sigCacheSize; i++ {
+		msg, sig, key, err := genRandomSig()
+		if err != nil {
+			t.Fatalf("unable to generate random signature test data")
+		}
+
+		sigCache.Add(msg, sig, key)
+
+		sigCopy, _ := btcec.ParseSignature(sig.Serialize(), btcec.S256())
+		keyCopy, _ := btcec.ParsePubKey(key.SerializeCompressed(), btcec.S256())
+
+		if !sigCache.Exists(msg, sigCopy, keyCopy) {
+			t.Errorf("previously added item not found in signature" +
+				"cache")
+		}
+	}
+
+	// The sigcache should now have sigCacheSize entries within it.
+	if len(sigCache.validSigs) != sigCacheSize {
+		t.Fatalf("sigcache should now have %v entries, instead it has %v",
+			sigCacheSize, len(sigCache.validSigs))
+	}
+
+	// Add a new entry, this should cause eviction of a randomly chosen
+	// previously entry.
+	msgNew, sigNew, keyNew, err := genRandomSig()
+	if err != nil {
+		t.Fatalf("unable to generate random signature test data")
+	}
+	sigCache.Add(msgNew, sigNew, keyNew)
+
+	// The sigcache should still have sigCache entries.
+	if len(sigCache.validSigs) != sigCacheSize {
+		t.Fatalf("sigcache should now have %v entries, instead it has %v",
+			sigCacheSize, len(sigCache.validSigs))
+	}
+
+	sigNewCopy, _ := btcec.ParseSignature(sigNew.Serialize(), btcec.S256())
+	keyNewCopy, _ := btcec.ParsePubKey(keyNew.SerializeCompressed(), btcec.S256())
+
```

Comment:
Inconsistent blank line.  The copies are created after the comment in the code below.